codec-http2/src/main/java/io/netty/handler/codec/http2/WeightedFairQueueByteDistributor.java

```diff
@@ -0,0 +1,289 @@
+    public void allocationQuantum(int allocationQuantum) {
+        if (allocationQuantum <= 0) {
+            throw new IllegalArgumentException("allocationQuantum must be > 0");
+        }
+        this.allocationQuantum = allocationQuantum;
+    }
+
+    private int distribute(int maxBytes, Writer writer, State pState) throws Http2Exception {
+        if (pState.hasFrame) {
+            int nsent = min(maxBytes, pState.streamableBytes);
+            pState.write(nsent, writer);
+            connectionState.pseudoTime += nsent;
+            return nsent;
+        }
+
+        return distributeToChildren(maxBytes, writer, pState);
+    }
+
+    private int distributeToChildren(int maxBytes, Writer writer, State pState) throws Http2Exception {
+        long oldTotalQueuedWeights = pState.totalQueuedWeights;
+        State iState = pState.poll();
+        State iNextState = pState.peek();
+        try {
+            assert iNextState == null || iNextState.pseudoTime >= iState.pseudoTime : "iNextState.pseudoTime(" +
+                    iNextState.pseudoTime + ") < " + " iState.pseudoTime(" + iState.pseudoTime + ")";
+            int nsent = distribute(iNextState == null ? maxBytes :
+                                    min(maxBytes, (int) ((iNextState.pseudoTime - iState.pseudoTime) *
+                                             ((double) iNextState.stream.weight() / oldTotalQueuedWeights)
+                                             + allocationQuantum)),
+                               writer,
+                               iState);
+            iState.updatePseudoTime(nsent, oldTotalQueuedWeights);
+            return nsent;
+        } finally { // Do in finally to ensure the internal state is not corrupted if an exception is thrown.
+            // The offer operation is delayed until we unroll up the recursive stack, so we don't have to remove from
+            // the priority queue due to a write operation.
+            if (iState.hasFrameCountForTree != 0) {
+                pState.offer(iState);
+            }
+        }
+    }
+
+    private State state(Http2Stream stream) {
+        return stream.getProperty(stateKey);
+    }
+
+    /**
+     * For testing only!
+     */
+    int streamableBytes(Http2Stream stream) {
+        return state(stream).streamableBytes;
+    }
+
+    /**
+     * The remote flow control state for a single stream.
+     */
+    private final class State {
+        final Http2Stream stream;
+        private final Queue<State> queue;
+        int streamableBytes;
+        int hasFrameCountForTree;
+        long pseudoTime;
+        long totalQueuedWeights;
+        boolean hasFrame;
+        private boolean isQueued;
+
+        State(Http2Stream stream) {
+            this(stream, 2, connectionState.pseudoTime);
+        }
+
+        State(Http2Stream stream, int initialSize, long pseudoTime) {
+            this.stream = stream;
+            queue = new PriorityQueue<State>(initialSize, WFQ_COMPARATOR);
```

Comment:
@Scottmitch I wonder if we could get away with not removing immediately. For example, we could just make it a rule that when we go to write to a stream, we do not write if the parent has data (or recursively any ancestor?). Then the logic might be something like this:

1) Parent streamable bytes becomes > 0: do not remove children from the queue immediately. Instead wait until we process the child, where we see that the parent has data still so we remove the child but do not write to it.

2) Parent streamable bytes falls to 0: recursively iterate through children and add those with frames to send.

I'm sure there are probably some edge cases to sort out, but that's the basic idea.  WDYT?

@ejona86 thoughts?